### ISSUE TITLE
training problem

### ISSUE THREAD

**Issue by kiki800917** [michaeltrs/DeepSatModels#6]
Thanks for your reply on last 2 issues, it is really helpful.
I implemented your model with temporal embedding first and spatial embedding second. with hyper params that mostly follow your best hyper params: 
task: semantic segmentation
patch size : 2x2
time series img size : [channels, num_frames, height, width] = [26, 5, 24, 24]
positional encoding:transformer encodings
num_training_samples = 6k
num_classes = 4
batch_size : 16/32/48 (I have tried all of them)
embed_dim = 128
heads = 4
loss_function = weighted cross entropy(to deal with data imbalance)
optimizer : Adam/AdamW
lr = 1e-3~1e-6
depth = 4(on both temporal and spatial encoder)
with no regularization

I have check my code twice and seems no bugs inside. But the model fail to converge in training data(underfitting), the loss is starting from 1.38 and best reduce to 0.89. But I used the simplest U-Net, it can converge to loss = 0.2(just use origin 351x351 img without clip then to 24x24, and no time series info given. a training data is a 351x351x26(HxWxC) img with a 351x351 label(4 classes)).I also found that most people fails to converge when using ViT(or other transformer based model on CV) to train their own dataset : https://github.com/lucidrains/vit-pytorch/issues/45
is it has some important thing that i miss? 

If possible, could you provide the training and valdation loss curves of classification and segmentation task on 3 different datasets you experiment?

My dataset is private, and there is no other using this dataset on any deep learning model before.This dataset are spectral images (shape:351x351x26(HxWxC)) with label which is a segmentation mask (shape = 351x351), 4 classes(no rain, small rain, medium, large) of rainfall classification. the spectral image is produced per 10 minutes, so i can make them be a time series images with shape [channels, timesteps, height, width]. And I have about 50k samples of shape [26, 5, 24, 24] = [channels, timesteps, height, width] cropped from the original spectral images.

**Comment by michaeltrs** [comment]
Hi @kiki800917,

About the difficulty of training ViT, this hasn't been our experience, most temp-spatial models shown in the paper typically behaved well from random initialisation without any tricks. I believe the difficulty of training in natural images has to do with the fact that transformers do not have the structural biases of CNNs that make them good for grid like data like images. As a result ViT needs smarter initialisations and longer training. In our case TSViT was designed with structural biases inherent in SITS for crop type recognition in mind.

However, these might not be the ideal choices for your specific problem. Initially, if your problem can be learned from a single image, then temporal features may not be crucial, and you might not require time-series modeling. As TSViT emphasizes temporal features, this could explain why the training loss isn't decreasing.

You also need to think about how to define the temporal position encodings. In [self.to_temporal_embedding_input = nn.Linear(366, self.dim)
](https://github.com/michaeltrs/DeepSatModels/blob/bb5ef2a8f4b5a2b3951de5e1a88647736308b330/models/TSViT/TSViTdense.py#LL397C1-L398C1) we define the size of the temporal encoding table to equal the number of days in a year because we use the day of the year as the temporal granularity, and samples can have different dates among them. If I understand correctly that your frames are consistently separated by time and you only need to encode their order, static position encodings should suffice. If you want to encode more information than order you need to modify self.to_temporal_embedding_input. If for example each 5 consecutive images were captured in different times during the day and that is important, you could initialise an encoding for every 10min interval in a day, change 366 to 144 (number of 10min intervals in a day) and index accordingly to get the correct temporal encodings.

The final point concerns the usage of multiple tokens. In our paper, we argued that identifying a crop type in one location doesn't constrain or provide useful information about the crop type at another location. Therefore, the spatial encoder processes tokens for each class separately. Recognizing rainfall patterns, however, may require consideration of these interrelationships. For instance, if you detect water at location x at time t, it's plausible to expect water nearby at time t+1.

In summary, TSViT might not be optimal for your problem. I would recommend the following steps if you want to investigate further: 1) Train a UNET on small 24x24x26 patches to examine the effect of limiting the spatial context; 2) Try a TSViT with static temporal encodings and a single token to investigate the impact of these modifications.

I hope this helps!
